### ISSUE TITLE
Tables: Semantics for better accessibility

### ISSUE THREAD

**Issue by toddparker** [COVID19Tracking/website#724]
This is part of the larger table work in #457

I'd like to review our current table semantics and make sure we're following best practices:
- [WAI tables tutotrial](https://www.w3.org/WAI/tutorials/tables/)
- [Deque tables checklist for WCAG](https://dequeuniversity.com/checklists/web/tables)

## To review
We should pay special attention to grouped or spanned columns/rows

- [ ] [US Totals History](https://covidtracking.com/data/us-daily) (flat)
- [ ] [State history](https://covidtracking.com/data/state/alabama#historical) (flat)
- [ ] [State summary](https://covidtracking.com/data/state/new-york) (grouped)
- [ ] [Totals by state](https://covidtracking.com/data) (grouped)

**Comment by beep** [comment]
FWIW I based the current markup off [the WAI’s “irregular headers” pattern](https://www.w3.org/WAI/tutorials/tables/irregular/), but I’m honestly not sure how useful `colgroup` is. Only noting it in case it’s helpful!

**Comment by toddparker** [comment]
@beep yeah, seems `scope` and the `id/headers` can both be used, not sure if there are differences in how well each is suppler by screen readers? 
https://developer.mozilla.org/en-US/docs/Learn/HTML/Tables/Advanced#Using_column_and_row_headers

**Comment by tatianamac** [comment]
I would say we want to be mindful of using `headers` and `ids` and reduce our use of them as much as possible as many screenreaders (SRs) cannot handle them very well across the board. Same goes with `colspan` and such.

**Comment by toddparker** [comment]
@tatianamac and I were talking about the idea of flattening these two-level headers and I think we might want to flip this problem around so there is just a single header tow - no `colspan`s funny business.

Idea: should we just have the name of the grouping repeated in each header, like this:
```
<th><span>Hospitalized</span> Current</th>
<th><span>Hospitalized</span> Cumulative</th>

On a screen reader, it would read back each clearly as **"Hospitalized Current"** and **"Hospitalized Cumulative"**  so it super clear, if a wee bit verbose.

We could then do some CSS hackery to position the first span up so it visually looks like a grouping and accessibly hide the rest of the `span`s in that group so it looks like the design.

Anyone with better CSS chops than me want to prototype that idea?

![image](https://user-images.githubusercontent.com/359929/80413685-d7346680-889d-11ea-9e06-f2f0397efc13.png)

**Comment by beep** [comment]
> Idea: should we just have the name of the grouping repeated in each header, like this:
> 
> ```
> <th><span>Hospitalized</span> Current</th>
> <th><span>Hospitalized</span> Cumulative</th>```

Oh, _interesting._ Lemme take a crack at this?

(I’m gonna see about putting the “grouping” at the end of the cell, since (e.g.) `Current Hospitalized` probably makes a bit more sense than `Hospitalized Current`—is that cool?)

**Comment by tatianamac** [comment]
@beep Yay! That makes sense to me, esp because it is true for all headers (that I see here), such as "Positive Tests" rather than "Tests, positive"

**Comment by toddparker** [comment]
@beep that's a great point. Seems that span should be wherever it needs to be to make the language read cleanly in each header. I agree here that should _after_ the more specific heading. Love to see what you cook up. cc: @tatianamac

**Comment by beep** [comment]
Okay, here’s a very quick but probably-overengineered sketch!

https://codepen.io/beep/pen/BaoZKoX

I didn’t muck with the styling specifics _too_ much, instead just looking at the header groupings. And on that note, the positioning feels like it’s…probably fine? With, y’know, the usual caveats about absolute positioning and wrapping, of course. I’d say it’s something to keep an eye on during prototyping.

The JavaScript should definitely be treated as FPO; it’s really only there to do two things:

- Flag the first instance of a given label as a candidate for styling, and then
- Accessibly hide all subsequent labels

(If this is gonna get Gatsby-ified, I’d recommend we handle this on the server.)

But! Was this roughly what you folks had in mind?

**Comment by schwartzadev** [comment]
Conceptually, this looks great!

For Gatsby, the way that we've been handling `sr-only` items is by making them hidden by default. It looks like you have the visibility the other way around here, but this is super minor -- we'll just have to make sure it gets pulled through into the code side.

**Comment by toddparker** [comment]
Dang @beep that seems to work like a charm! I think that will do the trick.

It's too bad there isn't some kind of fancy CSS selector for that to avoid JS but I guess since these are arbitrary `th` "groups" there isn't much to hang our hat on.

**Comment by toddparker** [comment]
@beep...just had a thought. Since we're going to need to add some extra classes on the cells to add in the left border for the column groupings, could we just add in the classes in the markup statically? Maybe we could make a version of this pen that adds in those grouping lines and just has these as static classes?

**Comment by beep** [comment]
@toddparker So! You sort of can! (This is kinda what I was over-engineering 😅)

Something like this would work just fine:

```
<table>
  <thead>
    <tr>
      <th data-group="tests">Positive <span class="label--header-group">Tests</span></th>
      <th data-group="tests">Negative <span class="label--header-group">Tests</span></th>
      <th data-group="tests">Pending <span class="label--header-group">Tests</span></th>
      <th data-group="tests">Total <span class="label--header-group">Tests</span></th>
      <th data-group="hosp">Current <span class="label--header-group">Hospitalized</span></th>
      <th data-group="hosp">Cumulative <span class="label--header-group">Hospitalized</span></th>
      <th data-group="icu">Current <span class="label--header-group">In ICU</span></th>
      <th data-group="icu">Cumulative <span class="label--header-group">In ICU</span></th>

…
.label--header-group {
  // Hide accessibly
}
[data-group]:first-child .label--header-group,
:not([data-group="hosp"]) + [data-group="hosp"] .label--header-group,
:not([data-group="icu"]) + [data-group="icu"] .label--header-group {
  // Embiggen the first label in the row, and then the first label in each “group”
}
```

This is a little fragile, as you might’ve guessed. The CSS would need to know about the groupings on the page, and then we’d need to build selectors for each. (It’s possible I’m missing a more elegant CSS-only approach, but it’s EOD and I’m a bit wiped.)

So—unless I’m missing something!—we’d need to get a lot more verbose with our markup to say “this is the first header in this group`, `this is the last header in this group`, `this is the first body cell in this group`, &c. And we can (as I did) employ a little light JS to help, or do that directly in the markup on the server. (yes let’s do that yes yes yes)

**Comment by beep** [comment]
> Maybe we could make a version of this pen that adds in those grouping lines and just has these as static classes?

Sure! I’ll do that in the morning.

**Comment by beep** [comment]
> For Gatsby, the way that we've been handling sr-only items is by making them hidden by default. It looks like you have the visibility the other way around here, but this is super minor -- we'll just have to make sure it gets pulled through into the code side.

@schwartzadev Could you say a bit more about this? (Happy to redirect this thread to Slack!)

**Comment by toddparker** [comment]
Oooh, good stuff @beep. I guess my next question is how we're going to add the border on each cell in each row. In the past we just ended up putting a class on the right cells but maybe there is a smart way to do that?

**Comment by beep** [comment]
@toddparker Sadly, I think a `class` is our best bet here. Gonna make for some chonky markup, but it’s gonna _look_ a treat 💅

**Comment by tealtan** [comment]
Looks like this is going well, thank you all! 👏

[Another recommendation](https://www.w3.org/WAI/tutorials/tables/caption-summary/#nesting-summary-inside-the-caption-element) about adding summaries for more complex tables seems useful here in complementing the table header reworking above.

**Comment by beep** [comment]
_Whoops_, forgot to post my updated pen:

https://codepen.io/beep/pen/BaoZKoX

@toddparker This roughly what you had in mind?

**Comment by toddparker** [comment]
@beep this looks amazing, I think this will do the tricky nicely! Any concerns with this approach?

I definitely get that because of the positioning, this is a wee bit fragile:
![image](https://user-images.githubusercontent.com/359929/80605711-6c9c3b80-8a01-11ea-80cc-dd12d25fe15d.png)

...seems we just need to be careful about header length or set some min widths if things get bad. Any suggestions there on making it less fool proof?

Minor thing but mind making the vertical dividers 1px instead of 2px? Just don't want to confuse anyone that the design changed.

**Comment by beep** [comment]
> Any suggestions there on making it more fool proof?

Thinking about this a bit:

I don’t think there’s a CSS-only solution that’d be really resilient, no. The _most_ foolproof solution would be a row of actual, real headers, which `colspan` as needed.

@tatianamac, a question for ya when you next get a minute: do you have any info/articles on the AT issues around non-flattened tables? Asking because a.) I’d love to understand it better, but _also_ b.) I wonder if we could, like, create an actual top-level header row for our cell groupings (as we currently have) that’s `aria-hidden` so it doesn’t donk up assistive tech.

> Minor thing but mind making the vertical dividers 1px instead of 2px? Just don't want to confuse anyone that the design changed.

Oops, sorry—yes! Updated:

https://codepen.io/beep/pen/BaoZKoX

**Comment by toddparker** [comment]
Thanks @beep! Re: headers, I could see that we could maybe use `aria-hidden` on the grouped header row, but yeah, I'm curious is that gets real odd on SRs because those colspans are still sort of in play? Reminds me that visual parity is important, dunno which of these is better for both of these considerations.

**Comment by tatianamac** [comment]
The most comprehensive list of resources on table accessibility is here:
[Adrian Roselli's collection, "Hey It's okay to use tables"](https://adrianroselli.com/2017/11/hey-its-still-ok-to-use-tables.html
).

A small note is in [WebAIM's guide](https://webaim.org/techniques/tables/data) as well.

**Comment by beep** [comment]
After some back-and-forth, and some gracious JAWS/NVDA testing by a friend of @tatianamac’s, here’s the most usable version of the table design we’d like to start implementing:

https://codepen.io/beep/pen/BaoZKoX

A few notes for implementation:

- Feel free to revisit the class names and styles as you see fit! Everything in the CodePen can be considered FPO, if you’d like.
- The left-hand borders at the start of each “group” are assigned via the `cell--group-start` class, applied to each `th` and `td` in that “group.”
- Each `th` follows this general structure:

```
<th scope="col" class="cell--group-start">
  <span role="text">
    Positive <span class="label--header-group">Tests</span>
  </span>
</th>
```

That is,

1. A `scope="col"` attribute, with an _optional_ `class="cell--group-start"` if the header’s denotes the start of the group.
1. A `span` with a `role="text"`, that wraps the header cell’s contents. ([Here’s why!](https://axesslab.com/text-splitting/))
1. The text of the header cell, which contains:
  1. The visible text of the header’s label (above, “Positive”);
  2. The visually hidden part of the label (above, “Tests”), wrapped in a `<span class="label--header-group">…</span>` element.

Basically, the `label--header-group` element is visually hidden _unless_ it’s contained in a `cell--group-start`-labeled cell; if it is, then the “group label” is positioned up above, and made to look as though it’s spanning the additional rows.

To whomever ends up implementing this in Gatsbyland (@dasulit?), the challenge during implementation is going to be avoiding any overlap of these positioned labels. I think there’s a lot you’ll be able to do with CSS here—just off the top of my head: reducing font size; making all the group labels appear inline; &c—but if some JS fanciness needs to be employed, you might have to employ a resizeObserver and do a little honest-to-goodness hit detection. (:lolsob:) Let me know if I can help brainstorm any of this.
